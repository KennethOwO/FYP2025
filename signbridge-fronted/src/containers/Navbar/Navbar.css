.navbar {
    background: #1c2e4a;
    height: 80px;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 1.2rem;
    position: sticky;
    top: 0;
    z-index: 999;
}

.navbar-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 80px;
    width: 100%;
}

.navbar-logo {
    justify-self: start;
    cursor: pointer;
    text-decoration: none;
    display: flex;
    align-items: center;
}

.navbar-logo-image {
    width: 100%;
    height: 50px;
}

.nav-menu {
    display: grid;
    grid-template-columns: repeat(7, auto);
    grid-gap: 40px;
    list-style: none;
    text-align: center;
    width: 65vw;
    justify-content: end;
    margin-right: 2rem;
    font-weight: 600;
}

.nav-item {
    height: 80px;
    position: relative;
}

.nav-item:hover .dropdown-menu {
    display: block;
    min-width: 110px;
}

.nav-links {
    color: #fff;
    display: flex;
    align-items: center;
    text-decoration: none;
    padding: 1.8rem 1rem 0.5rem 1rem;
    white-space: nowrap;
}

.nav-links:hover {
    border-bottom: 4px solid #b6b6b4;
    transition: all 0.2s ease-out;
}

.nav-item:hover .nav-links {
    border-bottom: 4px solid #ffffff;
    transition: all 0.2s ease-out;
}

.nav-links .fa-bell {
    margin-top: -5px;
    padding: 2px;
    font-size: 2rem;
}

.fa-bars {
    color: #fff;
}

.nav-links-mobile {
    display: none;
}

.menu-icon {
    display: none;
}

.dropdown-menu {
    display: none;
    position: absolute;
    background-color: #242222;
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
    z-index: 1;
    width: fit-content;
}

.dropdown-menu li {
    display: block;
    list-style: none;
}

.dropdown-menu li a {
    color: #fff;
    padding: 12px;
    text-decoration: none;
    display: block;
}

.dropdown-menu li a:hover {
    background-color: #111;
}

/* For Dataset Collection */
.dropdown-menu2 {
    display: none;
    position: absolute;
    background-color: #242222;
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
    z-index: 1;
    width: 130%;
}

.dropdown:hover .dropdown-menu2 {
    display: block;
    position: flex;
    margin-left: -30px;
}

.dropdown-menu2 li {
    display: block;
    list-style: none;
}

.dropdown-menu2 li a {
    color: #fff;
    padding: 12px 16px;
    text-decoration: none;
    display: block;
}

.dropdown-menu2 li a:hover {
    background-color: #111;
}

/* Bilingual */
.dropdown-menu3 {
    display: none;
    position: absolute;
    background-color: #242222;
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
    z-index: 1;
    width: 130%;
}

.dropdown-menu3 li {
    display: block;
    list-style: none;
}

.dropdown-menu3 li a {
    color: #fff;
    padding: 12px 16px;
    text-decoration: none;
    display: block;
}

.dropdown-menu3 li a:hover {
    background-color: #111;
}

.profileimg {
    color: #fff;
    margin-top: 15px;
    width: 50px;
    height: 50px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    text-decoration: none;
    object-fit: cover;
    object-position: center;
}

.profileimg:hover {
    border: 3px solid #fff;
    transition: all 0.2s ease-out;
}

.nav-item:hover .profileimg {
    border-bottom: 3px solid #ffffff;
    transition: all 0.2s ease-out;
}

.btn-mobile {
    margin: auto 0;
}

.notification-icon-container {
    position: relative;
    display: inline-block;
}

.notification-count {
    position: absolute;
    top: -5px;
    right: -8px;
    background-color: red;
    color: white;
    border-radius: 50%;
    padding: 3px 6px;
    font-size: 12px;
}

@media (width > 1300px) {
    .dropdown:hover .dropdown-menu {
        display: block;
        margin-left: -30px;
    }

    .dropdown-trigger-checkbox,
    .dropdown-trigger-checkbox2 {
        display: none;
    }

    .dropdown:hover .dropdown-menu3 {
        display: block;
        position: flex;
        margin-left: -9px;
    }
}

@media (width <= 1300px) {
    .dropdown-menu,
    .dropdown-menu2,
    .dropdown-menu3 {
        display: none;
    }

    .dropdown-trigger-checkbox,
    .dropdown-trigger-checkbox2,
    .dropdown-trigger-checkbox-dataset {
        display: hidden;
    }

    .dropdown-trigger-checkbox[type="checkbox"] + .dropdown-menu,
    .dropdown-trigger-checkbox2[type="checkbox"] + .dropdown-menu3,
    .dropdown-trigger-checkbox-dataset[type="checkbox"] + .dropdown-menu2 {
        display: none;
    }

    .dropdown-trigger-checkbox[type="checkbox"]:checked + .dropdown-menu,
    .dropdown-trigger-checkbox2[type="checkbox"]:checked + .dropdown-menu3,
    .dropdown-trigger-checkbox-dataset[type="checkbox"]:checked + .dropdown-menu2 {
        display: block;
        margin: auto 0;
    }

    .navbar {
        font-size: 1rem;
    }

    .nav-menu {
        display: flex;
        flex-direction: column;
        width: 100%;
        height: 95vh;
        position: absolute;
        top: 80px;
        left: -100%;
        opacity: 1;
        justify-content: start;
        transition: all 0.5s ease;
        overflow-y: auto;
    }

    .nav-menu.activate {
        background: #242222;
        left: 0;
        opacity: 1;
        transition: all 0.5s ease;
        z-index: 1;
    }

    .nav-item {
        margin: 10px 0 -50px 0;
    }

    .nav-links {
        text-align: center;
        padding: 0.8rem;
        width: 100%;
        display: table;
    }

    .nav-links:hover {
        background-color: #fff;
        color: #242424;
        border-radius: 0;
    }

    .navbar-logo {
        justify-self: start;
        height: 80px;
        width: fit-content;
    }

    .menu-icon {
        display: block;
        position: absolute;
        top: 0;
        right: 0;
        transform: translate(-100%, 60%);
        font-size: 1.8rem;
        cursor: pointer;
    }

    .fa-times {
        color: #fff;
        font-size: 2rem;
    }

    .nav-links-mobile {
        display: block;
        text-align: center;
        margin: auto;
        /* border-radius: 4px; */
        /* width: 80%; */
        text-decoration: none;
        font-size: 0.8rem;
        background-color: transparent;
        color: #fff;
        padding: 14px 20px;
        border: 1px solid #fff;
        transition: all 0.3s ease-out;
    }

    .nav-links-mobile:hover {
        background: #fff;
        color: #242424;
        transition: 250ms;
    }

    .nav-links .fa-bell {
        margin-top: -5px;
        padding: 2px;
        font-size: 1.2rem; /* Adjust the size as needed */
    }

    .dropdown-menu,
    .dropdown-menu2,
    .dropdown-menu3 {
        display: none;
        position: flex;
        background-color: #605e5e;
        box-shadow: 0 8px 16px 0 rgba(240, 240, 240, 0.2);
        z-index: 1;
        width: 100%;
    }

    /* .dropdown:hover .dropdown-menu,
    .dropdown:hover .dropdown-menu2,
    .dropdown:hover .dropdown-menu3 {
        display: block;
        position: flex;
        margin: auto 0;
    } */

    .profileimg {
        margin: auto;
        width: 40px;
        height: 40px;
    }
}
