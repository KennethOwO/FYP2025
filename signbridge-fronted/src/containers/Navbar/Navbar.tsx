import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Button } from "../../components/Button/Button";
import { useThemeStore } from "../../store/theme";
import "./Navbar.css";
import { StreamNotifications } from "../../services/notification.service";
import { Globe } from "lucide-react";
import { useTranslation } from "react-i18next";
import { auth } from "../../../firebase";
import { getAuth, signOut } from "firebase/auth";
import toast from "react-hot-toast";

import { useUserStore } from "@root/store/userStore";

function Navbar() {
    const navigate = useNavigate();
    const [click, setClick] = useState(false);
    const [button, setButton] = useState(true);

    const { user } = useUserStore();
    const { color, updateColors } = useThemeStore();

    const { t, i18n } = useTranslation();
    const currentUser = getAuth().currentUser;
    const changeLanguage = (lang: string) => {
        i18n.changeLanguage(lang);
        localStorage.setItem("i18nextLng", lang);
        closeMobileMenu();
    };

    useEffect(() => {
        const currentLang = localStorage.getItem("i18nextLng") || "en"; // Default to English
        i18n.changeLanguage(currentLang);
    }, [i18n]);

    // Function to clear user authentication status from session storage when user logs out
    const handleLogout = async () => {
        await signOut(auth).then(() => {
            localStorage.setItem("color", "#1C2E4A");
            updateColors("#1C2E4A");
            toast.success(t("logout_successful"));
            navigate("/login");
        });
    };

    const handleClick = () => setClick(!click);
    const closeMobileMenu = () => setClick(false);

    // ---------- Function to show login button on small screens ----------
    const showButton = () => {
        if (window.innerWidth <= 1300) {
            setButton(false);
        } else {
            setButton(true);
        }
    };

    useEffect(() => {
        showButton();
    }, []);

    // ---------- Fetch notification counts ----------
    const [notificationCount, setNotificationCount] = useState(0);

    useEffect(() => {
        let eventSourceCleanup: (() => void) | undefined;

        const setupNotificationStream = async () => {
            if (user?.user_id) {
                eventSourceCleanup = await StreamNotifications(user.user_id, count => {
                    setNotificationCount(count);
                });
            }
        };

        setupNotificationStream();

        return () => {
            if (eventSourceCleanup) {
                eventSourceCleanup();
            }
        };
    }, [user?.user_id]);

    // ---------- Listen on the weebsite screen size ----------
    window.addEventListener("resize", showButton);

    return (
        <>
            <nav className="navbar" style={{ background: color }}>
                <div className="navbar-container">
                    {/* ---------- Webiste's logo ---------- */}
                    <Link to="/" className="navbar-logo" onClick={closeMobileMenu}>
                        <img src="/images/PUBLIC_LOGO.png" alt="Logo" className="navbar-logo-image" />
                    </Link>

                    {/* ---------- Hamburger menu icon (Small screen size) ---------- */}
                    <div className="menu-icon" onClick={handleClick}>
                        <i className={click ? "fas fa-times" : "fas fa-bars"} />
                    </div>

                    {/* ---------- Navbar menu ---------- */}
                    <ul className={click ? "nav-menu activate" : "nav-menu"}>
                        <li className="nav-item dropdown">
                            <label htmlFor="toggle1" className="nav-links" style={{ cursor: "pointer" }}>
                                {t("module")}
                            </label>
                            <input type="checkbox" id="toggle1" name="toggle" className="dropdown-trigger-checkbox" hidden />

                            <ul className="dropdown-menu">
                                <li>
                                    <Link to="/library" className="dropdown-link" onClick={closeMobileMenu}>
                                        {t("library")}
                                    </Link>
                                </li>

                                <li>
                                    <Link to="/communication" className="dropdown-link" onClick={closeMobileMenu}>
                                        {t("communication")}
                                    </Link>
                                </li>

                                <li>
                                    <Link to="/education" className="dropdown-link" onClick={closeMobileMenu}>
                                        {t("education")}
                                    </Link>
                                </li>
                            </ul>
                        </li>
                        {user?.role_access === "signexpert" ? (
                            <li className="nav-item dropdown">
                                <label htmlFor="toggle-dataset" className="nav-links" style={{ cursor: "pointer" }}>
                                    {t("dataset_collection")}
                                </label>
                                <input type="checkbox" id="toggle-dataset" name="toggle-dataset" className="dropdown-trigger-checkbox-dataset" hidden />
                                <ul className="dropdown-menu2">
                                    <li>
                                        <Link to="/dataset-collection" className="dropdown-link" onClick={closeMobileMenu}>
                                            {t("dataset_form")}
                                        </Link>
                                    </li>
                                    <li>
                                        <Link to="/dataset-collection-review" className="dropdown-link" onClick={closeMobileMenu}>
                                            {t("dataset_review")}
                                        </Link>
                                    </li>
                                </ul>
                            </li>
                        ) : (
                            <li className="nav-item">
                                <Link to="/dataset-collection" className="nav-links" onClick={closeMobileMenu}>
                                    {t("dataset_collection")}
                                </Link>
                            </li>
                        )}

                        <li className="nav-item">
                            <Link to="/feedback" className="nav-links" onClick={closeMobileMenu}>
                                {t("feedback")}
                            </Link>
                        </li>

                        <li className="nav-item">
                            <Link to="/faq" className="nav-links" onClick={closeMobileMenu}>
                                {t("faq")}
                            </Link>
                        </li>

                        <li className="nav-item dropdown">
                            <label className="nav-links" htmlFor="toggle2" style={{ cursor: "pointer" }}>
                                <Globe />
                            </label>
                            <input type="checkbox" id="toggle2" name="toggle2" className="dropdown-trigger-checkbox2" hidden />
                            <ul className="dropdown-menu3">
                                <li>
                                    <Link to="#" className="dropdown-link" onClick={() => changeLanguage("en")}>
                                        EN
                                    </Link>
                                </li>

                                <li>
                                    <Link to="#" className="dropdown-link" onClick={() => changeLanguage("bm")}>
                                        BM
                                    </Link>
                                </li>
                            </ul>
                        </li>

                        {!user ? (
                            // Render login button for users who are not logged in
                            <>
                                <div className="nav-item">
                                    <Link to="/login" className="nav-links-mobile" onClick={closeMobileMenu}>
                                        {t("login")}
                                    </Link>
                                </div>
                                {/* ---------- Login button (Small screen size) ---------- */}
                                <Link to="/login" className="btn-mobile">
                                    {button && (
                                        <Button type="button" onClick={closeMobileMenu} buttonStyle="btn--outline" buttonSize="btn--mobile">
                                            {t("login")}
                                        </Button>
                                    )}
                                </Link>
                            </>
                        ) : (
                            // Render profile and notifications icons for logged-in users
                            <>
                                <li className="nav-item">
                                    <Link to="/notifications" className="nav-links notification" onClick={closeMobileMenu}>
                                        <div className="notification-icon-container">
                                            <i className="fas fa-bell" />
                                            {notificationCount > 0 && <span className="notification-count">{notificationCount}</span>}
                                        </div>
                                    </Link>
                                </li>

                                <li className="nav-item dropdown">
                                    <img src={user?.picture} referrerPolicy="no-referrer" alt="Profile" className="profileimg" />

                                    <ul className="dropdown-menu">
                                        <li>
                                            <Link to="/profile" className="dropdown-link" onClick={closeMobileMenu}>
                                                {user?.username}
                                            </Link>
                                        </li>
                                        <li>
                                            <Link to="/login" className="dropdown-link" onClick={handleLogout}>
                                                {t("logout")}
                                            </Link>
                                        </li>
                                    </ul>
                                </li>
                            </>
                        )}
                    </ul>
                </div>
            </nav>
        </>
    );
}

export default Navbar;
