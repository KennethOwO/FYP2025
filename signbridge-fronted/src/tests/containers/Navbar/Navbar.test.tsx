import Navbar from "@root/containers/Navbar/Navbar";
import { create } from "react-test-renderer";
import { MemoryRouter } from "react-router-dom"; // Import MemoryRouter
import { cleanup } from "@testing-library/react";
jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key, // Mock the translation function to return the key
    i18n: {
      changeLanguage: jest.fn(), // Mock the changeLanguage function
    },
  }),
}));
describe("Test Navbar", () => {
  afterEach(() => {
    // Clean up the DOM
    cleanup();
    // Reset mocks to their initial state
    jest.resetAllMocks();
  });

  it("should render correctly", () => {
    const tree = create(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
