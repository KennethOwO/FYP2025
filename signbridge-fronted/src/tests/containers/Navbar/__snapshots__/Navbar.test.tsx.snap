// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Navbar should render correctly 1`] = `
<nav
  className="navbar"
  style={
    {
      "background": "#1C2E4A",
    }
  }
>
  <div
    className="navbar-container"
  >
    <a
      className="navbar-logo"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="Logo"
        className="navbar-logo-image"
        src="./images/PUBLIC_LOGO.png"
      />
    </a>
    <div
      className="menu-icon"
      onClick={[Function]}
    >
      <i
        className="fas fa-bars"
      />
    </div>
    <ul
      className="nav-menu"
    >
      <li
        className="nav-item dropdown"
      >
        <a
          className="nav-links"
          href="/"
          onClick={[Function]}
        >
          module
        </a>
        <ul
          className="dropdown-menu"
        >
          <li>
            <a
              className="dropdown-link"
              href="/library"
              onClick={[Function]}
            >
              library
            </a>
          </li>
          <li>
            <a
              className="dropdown-link"
              href="/communication"
              onClick={[Function]}
            >
              communication
            </a>
          </li>
          <li>
            <a
              className="dropdown-link"
              href="/education"
              onClick={[Function]}
            >
              education
            </a>
          </li>
        </ul>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-links"
          href="/dataset-collection"
          onClick={[Function]}
        >
          dataset_collection
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-links"
          href="/feedback"
          onClick={[Function]}
        >
          feedback
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-links"
          href="/faq"
          onClick={[Function]}
        >
          faq
        </a>
      </li>
      <li
        className="nav-item dropdown"
      >
        <a
          className="nav-links"
          href="/"
          onClick={[Function]}
        >
          <svg
            className="lucide lucide-globe"
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="M12 2a14.5 14.5 0 0 0 0 20 14.5 14.5 0 0 0 0-20"
            />
            <path
              d="M2 12h20"
            />
          </svg>
        </a>
        <ul
          className="dropdown-menu3"
        >
          <li>
            <a
              className="dropdown-link"
              href="/"
              onClick={[Function]}
            >
              EN
            </a>
          </li>
          <li>
            <a
              className="dropdown-link"
              href="/"
              onClick={[Function]}
            >
              BM
            </a>
          </li>
        </ul>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-links-mobile"
          href="/login"
          onClick={[Function]}
        >
          login
        </a>
      </li>
      <a
        className="btn-mobile"
        href="/login"
        onClick={[Function]}
      >
        <button
          className="btn btn--outline btn--medium"
          data-testid="button-btn--outline"
          onClick={[Function]}
          type="button"
        >
          login
        </button>
      </a>
    </ul>
  </div>
</nav>
`;
